### ISSUE TITLE
Scheduler throughput reduced when many gated pods

### ISSUE THREAD

**Issue by gabesaba** [kubernetes/kubernetes#124384]
### What happened?

When handling a delete pod event, we attempt to remove pods from the unscheduablePods pool [1], taking a lock while doing so [2]. Gated pods reside in this unscheduable pool, and we process all of them for each delete event. Scheduling throughput is affected, likely because ScheduleOne also takes a lock [3]

[1] https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/eventhandlers.go#L244-L249

[2] https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/internal/queue/scheduling_queue.go#L1093-L1097

[3] https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/schedule_one.go#L69

### What did you expect to happen?

Scheduling throughput should not reduce when there are many gated pods. We eventually decide against moving these pods, but after some processing. https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/scheduler/internal/queue/scheduling_queue.go#L1178-L1183

### How can we reproduce it (as minimally and precisely as possible)?

Create many (~10000) gated pods. Then, create non-gated pods which eventually terminate. As these pods finish and pod delete events are processed, scheduling throughput decreases

### Kubernetes version

<details>

```console
$ kubectl version
Server Version: v1.29.1-gke.1589017
```
</details>

### Cloud provider

<details>
Google Cloud - GKE
</details>

**Comment by gabesaba** [comment]
cc @alculquicondor

**Comment by alculquicondor** [comment]
@Huang-Wei, do you remember what the justification was to put the gated pods in the same unschedulable pool of pods?

I think we should just keep them entirely separate, so that gated pods truly don't affect scheduling throughput.

**Comment by sanposhiho** [comment]
I don't think we don't need to take a lock before checking `isPodWorthRequeuing`. 
Can we just refactor `movePodsToActiveOrBackoffQueue` so that it takes the lock after `isPodWorthRequeuing` is done, rather than requiring lock before `movePodsToActiveOrBackoffQueue` is called?

**Comment by chengjoey** [comment]
I tried to create 10,000 gated-pods, and then whether I changed the gated-pods to non-gated-pods or created normal pods, I did not find that the throughput was reduced. Is there something I am missing? I think the scheduling is well.

![image](https://github.com/kubernetes/kubernetes/assets/30427474/cb584364-d074-4ad6-b4be-5b12a58987b9)
![image](https://github.com/kubernetes/kubernetes/assets/30427474/d0f68c51-361c-4d7e-b8d9-ea79f7e4ae15)

**Comment by Huang-Wei** [comment]
> @Huang-Wei, do you remember what the justification was to put the gated pods in the same unschedulable pool of pods?

It's mainly for keeping the queueing code maintainable. Given the transition among activeQ/backoffQ/unschedulableQ is already complex, adding another internal queue doesn't sound a good idea by that time.

> I think we should just keep them entirely separate, so that gated pods truly don't affect scheduling throughput.

I doubt if throughput is really impacted... In terms of the re-queuing check on gated pods, (I haven't went through the latest code w/ schedulingHint), the original version is fairly straightforward - a PodDeletion event won't interest the gated pod at all, so it simply return. Even if the throughput is degraded, it's suspicious that the dominant factor is iterating gated pods. Needless to say, all re-queueing check happens in a separate goroutine.

So if throughput is proven impacted, I'd suggest to narrow down the real problem; otherwise moving gated pods to a separate struct won't benefit that much.

> Can we just refactor movePodsToActiveOrBackoffQueue so that it takes the lock after isPodWorthRequeuing is done, rather than requiring lock before movePodsToActiveOrBackoffQueue is called?

We should pin-point the specific issue by crafting a reproducible test, and then work on the solution. Maybe it's lock, maybe it's other things.

**Comment by gabesaba** [comment]
Here is a test I ran which reproduces the issue:

Create 10k gated pods
Create 30k pods at rate of 50/s, which sleep for 5 minutes each

Observe pods scheduling at ~50/s for first 5 minutes. When pods start terminating, the scheduler throughput decreases, and a backlog of pods builds up

![gated_performance_10k](https://github.com/kubernetes/kubernetes/assets/15304068/d922494c-7bd5-49af-9e1a-a9abb7e3a83f)

**Comment by gabesaba** [comment]
60s pprof sample with same experiment setup as comment directly above, after pods started terminating:
![pprof](https://github.com/kubernetes/kubernetes/assets/15304068/46d6edfb-666d-4f14-8862-30a3896faba7)

61.7% of samples are from deletePodFromCache, with 55.3% attributable to these 4 lines:
* 15.8% [scheduling_queue.go#L421](https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/internal/queue/scheduling_queue.go#L421)
* 14.5% [scheduling_queue.go#L434](https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/internal/queue/scheduling_queue.go#L434)
* 13.2% [scheduling_queue.go#L1149](https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/internal/queue/scheduling_queue.go#L1149)
* 11.8% [scheduling_queue.go#L443](https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/internal/queue/scheduling_queue.go#L443)

The smallest change would be to filter gated pods inline [here](https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/internal/queue/scheduling_queue.go#L1082). Could also use a preCheck function, but that is more than one line :)

/assign gabesaba

**Comment by sanposhiho** [comment]
First of all, this specific issue will be solved after [#122628](https://github.com/kubernetes/kubernetes/pull/122628); after [#122628](https://github.com/kubernetes/kubernetes/pull/122628), Pod deletion events won't reach `isPodWorthRequeuing`. 

But, the point is that this issue shows a **general** possibility where too many Pods in unschedQ can have a negative impact on scheduling cycles.
We can easily imagine a similar degradation happens when numerous Pods are in the unschedulable Pod pool, regardless of whether the reason Pods are in the unschedulable Pod pool is the gate or not.

So, I'd suggest moving `isPodWorthRequeuing` outside the lock (https://github.com/kubernetes/kubernetes/issues/124384#issuecomment-2065675227) as a general solution, rather than having a fix for this specific scenario.

**Comment by alculquicondor** [comment]
> The point is that this issue shows a general possibility where too many Pods in unschedQ can have a negative impact on scheduling cycles.

I agree. Just note that the motivation for scheduling gates is to be able to have a lot of gates pods without affecting the scheduling performance. So this issue shows a weakness of the current implementation.

OTOH, having a lot of pending pods has always been a problem.

> So, I'd suggest moving isPodWorthRequeuing outside the lock (https://github.com/kubernetes/kubernetes/issues/124384#issuecomment-2065675227) as a general solution, rather than having a fix for this specific scenario.

Are you sure that is possible? In order to run `isPodWorthRequeueing`, we need to obtain the list of pods from the queues, for which we need a lock. If we release the lock at this point, then the list might change and I'm not sure if we could run into more problematic race conditions.

**Comment by AxeZhan** [comment]
> Are you sure that is possible? In order to run isPodWorthRequeueing, we need to obtain the list of pods from the queues, for which we need a lock. If we release the lock at this point, then the list might change and I'm not sure if we could run into more problematic race conditions.

I think maybe what sanposhiho means is to release lock before `isPodWorthRequeueing` and acquire it again as soon as `isPodWorthRequeueing` finishes?

Maybe something like this:
```
p.lock()
podsToMove = p.unschedulablePods
p.unlock()

for po in podsToMove:
    schedulingHint := p.isPodWorthRequeuing
    p.lock()
    p.requeuePodViaQueueingHint
    p.unlock()
```

We only reading `queueingHintMap` during  `isPodWorthRequeueing` , so it should be fine(?). But I'm not sure if this will be very helpful.

**Comment by alculquicondor** [comment]
Right, I get that. My point is that it sounds very risky.

Definitely not a change that I would be comfortable backporting. And to make the change in master, I would like to see integration tests that exercise parallel pod deletions and creations. But I'm not really sure how the test should look like.

A fix targeting gated pods specifically can be much less risky.

**Comment by Huang-Wei** [comment]
> 61.7% of samples are from deletePodFromCache, with 55.3% attributable to these 4 lines:

It seems the major contributing factors are related w/ queueingHint's (core) bits. Which means although the feature gate is disabled, the core logic cannot be disabled.

Given it's not easy to switch back to the moment queueingHint core logic was not introduced, I agree with Aldo that handle gatedPods specifically is a portable fix.

@gabesaba FYI, the original (v1.27) impl. is [here](https://github.com/kubernetes/kubernetes/blob/96b450c75ae3c48037f651b4777646dcca855ed0/pkg/scheduler/internal/queue/scheduling_queue.go#L775C1-L785C4), which merely depends on a lightweight [podMatchesEvent()](https://github.com/kubernetes/kubernetes/blob/96b450c75ae3c48037f651b4777646dcca855ed0/pkg/scheduler/internal/queue/scheduling_queue.go#L1154) function.

**Comment by sanposhiho** [comment]
OK, so we're going to cherry-pick. Then I agree with a quick patch for gated pods proposed earlier.

**Comment by sanposhiho** [comment]
/assign

I'll explorer a general fix, including tests suggested by Aldo.

**Comment by AxeZhan** [comment]
So what's our conclusion now,

Make a quick fix with separate gated pods from unschedulable pool, and in the mean time, trying to find a general fix ?
If what I understand is correct, I can help with this quick fix while sanposhiho is finding a general fix.

**Comment by Huang-Wei** [comment]
@AxeZhan I think @gabesaba earlier self-assigned to explore a back-portable fix.

**Comment by AxeZhan** [comment]
> @AxeZhan I think @gabesaba earlier self-assigned to explore a back-portable fix.

Ah, didn't notice that, sry.

**Comment by Huang-Wei** [comment]
Btw: i think the title is a bit vague - throughout meant the ratio that main scheudling goroutine handles pods; while in this issue, it's more on the ratio of how scheduling queue moves pods. @alculquicondor @sanposhiho @gabesaba I'd suggest rewording the title to avoid confusion.

**Comment by gabesaba** [comment]
> Btw: i think the title is a bit vague - throughout meant the ratio that main scheudling goroutine handles pods; while in this issue, it's more on the ratio of how scheduling queue moves pods. @alculquicondor @sanposhiho @gabesaba I'd suggest rewording the title to avoid confusion.

Though this issue is impacting the rate at which we schedule pods - see the graphs in https://github.com/kubernetes/kubernetes/issues/124384#issuecomment-2066116284. Perhaps due to this call to Activate (which needs the SchedulingQueue's lock) in schedulingCycle: https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/schedule_one.go#L252-L257

**Comment by gabesaba** [comment]
We acquire the scheduling queue's lock in these two places - though the 2nd is probably irrelevant as it is done async

https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/schedule_one.go#L69

https://github.com/kubernetes/kubernetes/blob/00117569f3ab6456791c4e65fc4543a4278d1332/pkg/scheduler/schedule_one.go#L133

**Comment by sanposhiho** [comment]
From my view: 

This problem highlights two aspects:
- Scheduler's event handling takes a longer time than the previous version (which we don't have QHint).
- Scheduling latency is impacted negatively when event handling is taking time (event handling throughput is slow, there are too many incoming events, etc).

Based on them, action items:
- Decouple the scheduling throughput from event handling throughput. 
  - [ ] A quick fix: filter out gated Pods before `isPodWorthRequeuing`, which will fix the issue in this particular scenario. @gabesaba 
  - [ ] A general fix: moving `isPodWorthRequeuing` outside the lock, or maybe exploring another option. @sanposhiho 
- Observe and _maybe_ improve event handling throughput. After the decoupling above, scheduling throughput and event handling throughput would be two different metrics. (= I mean, one is supposed to have little impact on another.) We need to find how fast event handling should be on a certain large scale, like we have 300 pods/s target in scheduling throughput. And then we may or may not find that the event handling throughput needs to be improved accordingly.
  - [ ] 1. Add an observability for event handling throughput.
  - [ ] 2. Benchmark and decide the ideal throughput.
  - [ ] 3. Improve the throughput based on (2). (possible actions: simplify slower QHint(s),  etc)

(Probably I should create another separated issue for second one(s), rather putting too much stuff here.)

**Comment by alculquicondor** [comment]
> Scheduler's event handling takes a longer time than the previous version (which we don't have QHint).

TBH, we don't have proof of that. We didn't test 1.25. But it's still worth investigating how to improve the performance overall.

> 1. Add an observability for event handling throughput.
> 2. Benchmark and decide the ideal throughput.
> 3. Improve the throughput based on (2). (possible actions: simplify slower QHint(s), etc)

I agree with all of this. TBH, profiling seems to be the best tool for finding the hotspots. And then metrics can help us monitor in the future.

>  moving `isPodWorthRequeuing` outside the lock.

Maybe this is good step forward, and we could potentially guard it with the SchedulingHints feature gate.

**Comment by Huang-Wei** [comment]
> We need to find how fast event handling should be on a certain large scale, like we have 300 pods/s target in scheduling throughput

and also the duration each `moveAllToActiveOrBackoffQueue()` operation spent, and the number of getting called per sec/min.

**Comment by sanposhiho** [comment]
> TBH, we don't have proof of that. We didn't test 1.25. But it's still worth investigating how to improve the performance overall.

Ah, okay. I thought it was a degradation from the past version.
But, yeah, we can see the difference from the past based on the benchmark we will add.

> And then metrics can help us monitor in the future.

> Maybe this is good step forward, and we could potentially guard it with the SchedulingHints feature gate.

> and also the duration each moveAllToActiveOrBackoffQueue() operation spent

Agree with all.

**Comment by sanposhiho** [comment]
created: https://github.com/kubernetes/kubernetes/issues/124566

**Comment by alculquicondor** [comment]
Oh wait, was 1.28 the version that added hints for the first time? I got confused with this line https://github.com/kubernetes/enhancements/blob/3feb698013b4c227686a972e1da7e93e5fd0dcee/keps/sig-scheduling/4247-queueinghint/kep.yaml#L24

**Comment by gabesaba** [comment]
I reran the experiment from https://github.com/kubernetes/kubernetes/issues/124384#issuecomment-2066116284 with v1.27.11-gke.1062000, and indeed saw better performance (still observed some slowdown)

![experiment_127](https://github.com/kubernetes/kubernetes/assets/15304068/938dcd29-55cf-4eea-8b82-f8787a9022ce)

**Comment by sanposhiho** [comment]
/reopen

I'll work on a general fix mentioned in https://github.com/kubernetes/kubernetes/issues/124384#issuecomment-2076285212.

**Comment by sanposhiho** [comment]
/close

No more tracking is necessary from here.
Regarding the metrics stuff, we already have: https://github.com/kubernetes/kubernetes/issues/124566.

For the general solution discussed above, we had another discussion on https://github.com/kubernetes/kubernetes/issues/125223, and we actually have PRs for it.
- https://github.com/kubernetes/kubernetes/pull/125820
- https://github.com/kubernetes/kubernetes/pull/125692
